### PR TITLE
Add progress indicator to output logs

### LINKED ISSUE #223: progress indicator
It would be nice to have some idea how far along the running tests has gotten, given the multi-hour duration of conformance tests.
/kind feature
/sig testing

Right now there's _nothing_ logged until the test completes, and the spinner doesn't inspire confidence over multiple hours that we're actually getting anywhere.

cc @pohly for thoughts on how best to integrate with the ginkgo reporter.

My rough thinking: We should be able to have ginkgo at least print a "." per test and then count those and include them alongside the spinner, plus parse the "will run x out of y specs"

Something like:

```console
$ hydrophone --conformance
22:16:26 INF API endpoint: https://127.0.0.1:36061
22:16:26 INF Server version: version.Info{Major:"1", Minor:"32", GitVersion:"v1.32.0", GitCommit:"70d3cc986aa8221cd1dfb1121852688902d3bf53", GitTreeState:"clean", BuildDate:"2024-12-13T19:19:34Z", GoVersion:"go1.23.3", Compiler:"gc", Platform:"linux/amd64"}
22:16:26 INF Using namespace: conformance
22:16:26 INF Using conformance image: registry.k8s.io/conformance:v1.32.0
22:16:26 INF Using busybox image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
22:16:26 INF Test framework will start 1 thread(s) and use verbosity level 4.
22:16:26 INF Created Namespace conformance.
22:16:26 INF Created ServiceAccount conformance-serviceaccount.
22:16:26 INF Created Clusterrole conformance-serviceaccount:conformance.
22:16:26 INF Created ClusterRoleBinding conformance-serviceaccount-role:conformance.
22:16:26 INF Created Pod e2e-conformance-test.
22:16:26 INF Waiting up to 5m0s for Pod to start...
⠈⠁2024/12/16 22:16:29 Running command:
Command env: []
Run from directory: 
Executable path: /usr/local/bin/ginkgo
Args (comma-delimited): /usr/local/bin/ginkgo,--focus=\[Conformance\],--skip=,--no-color=true,--timeout=24h,/usr/local/bin/e2e.test,--,--disable-log-dump,--repo-root=/kubernetes,--provider=skeleton,--report-dir=/tmp/results,--kubeconfig=
2024/12/16 22:16:29 Now listening for interrupts
  I1216 22:16:30.364816      33 e2e.go:109] Starting e2e run "5e702b17-ac49-47b3-b5ac-fdede8df27dd" on Ginkgo node 1
Running Suite: Kubernetes e2e suite - /usr/local/bin
====================================================
Random Seed: 1734387389 - will randomize all specs

Will run 411 of 6622 specs
⠈⡱
```

Could instead be like:

```console
$ hydrophone --conformance
22:16:26 INF API endpoint: https://127.0.0.1:36061
22:16:26 INF Server version: version.Info{Major:"1", Minor:"32", GitVersion:"v1.32.0", GitCommit:"70d3cc986aa8221cd1dfb1121852688902d3bf53", GitTreeState:"clean", BuildDate:"2024-12-13T19:19:34Z", GoVersion:"go1.23.3", Compiler:"gc", Platform:"linux/amd64"}
22:16:26 INF Using namespace: conformance
22:16:26 INF Using conformance image: registry.k8s.io/conformance:v1.32.0
22:16:26 INF Using busybox image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
22:16:26 INF Test framework will start 1 thread(s) and use verbosity level 4.
22:16:26 INF Created Namespace conformance.
22:16:26 INF Created ServiceAccount conformance-serviceaccount.
22:16:26 INF Created Clusterrole conformance-serviceaccount:conformance.
22:16:26 INF Created ClusterRoleBinding conformance-serviceaccount-role:conformance.
22:16:26 INF Created Pod e2e-conformance-test.
22:16:26 INF Waiting up to 5m0s for Pod to start...
2024/12/16 22:16:29 Running command:
Command env: []
Run from directory: 
Executable path: /usr/local/bin/ginkgo
Args (comma-delimited): /usr/local/bin/ginkgo,--focus=\[Conformance\],--skip=,--no-color=true,--timeout=24h,/usr/local/bin/e2e.test,--,--disable-log-dump,--repo-root=/kubernetes,--provider=skeleton,--report-dir=/tmp/results,--kubeconfig=
2024/12/16 22:16:29 Now listening for interrupts
2024/12/16 22:16:30 Tests started with 411 out of 6622 to run

⢎⡠ [200/411] ...
```

### PR DESCRIPTION
**Pull request by farazkhawaja**
Implements progress tracking by calculating percentage based on processed vs total test count to output logs during test execution

Periodically reads e2e.log from conformance pod via k8s exec API every 30 seconds

Original inspiration from @rjsadow work on progress indicator -- https://github.com/kubernetes-sigs/hydrophone/pull/225

Resolves #223 
```

14:39:35 INF API endpoint: https://127.0.0.1:6443
14:39:35 INF Server version: version.Info{Major:"1", Minor:"31", GitVersion:"v1.31.5+k3s1", GitCommit:"56ec5dd4d012c1d77dc7f50f21f65183044c92e7", GitTreeState:"clean", BuildDate:"2025-01-28T17:45:21Z", GoVersion:"go1.22.10", Compiler:"gc", Platform:"linux/amd64"}
14:39:35 INF Using namespace: conformance
14:39:35 INF Using conformance image: registry.k8s.io/conformance:v1.31.5
14:39:35 INF Using busybox image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
14:39:35 INF Test framework will start 1 thread(s) and use verbosity level 4.
14:39:35 INF Created namespace conformance.
14:39:35 INF Created ServiceAccount conformance-serviceaccount.
14:39:35 INF Created ClusterRole conformance-serviceaccount:conformance.
14:39:35 INF Created ClusterRoleBinding conformance-serviceaccount-role:conformance.
14:39:35 INF Created Pod e2e-conformance-test.
14:39:35 INF Waiting up to 5m0s for Pod to start...
14:39:36 INF Created ConformancePod e2e-conformance-test.
2025/04/02 14:39:36 Running command:
Command env: []
Run from directory: 
Executable path: /usr/local/bin/ginkgo
Args (comma-delimited): /usr/local/bin/ginkgo,--focus=\[Conformance\],--skip=,--no-color=true,--timeout=24h,/usr/local/bin/e2e.test,--,--disable-log-dump,--repo-root=/kubernetes,--provider=skeleton,--report-dir=/tmp/results,--kubeconfig=
2025/04/02 14:39:36 Now listening for interrupts
  I0402 14:39:36.721627      24 e2e.go:109] Starting e2e run "68a87f29-c011-4b98-93bc-4e6ecc16f28e" on Ginkgo node 1
Running Suite: Kubernetes e2e suite - /usr/local/bin
====================================================
Random Seed: 1743604776 - will randomize all specs

Will run 404 of 6605 specs
[14:39:41] Progress: 1/404 tests completed (0.2%)
[14:40:11] Progress: 7/404 tests completed (1.7%)
[14:40:41] Progress: 11/404 tests completed (2.7%)
[14:41:11] Progress: 19/404 tests completed (4.7%)
```
Completed hour long conformance test 

```
[16:19:53] Progress: 389/404 tests completed (96.3%)
[16:20:23] Progress: 389/404 tests completed (96.3%)
[16:20:53] Progress: 389/404 tests completed (96.3%)
[16:21:23] Progress: 389/404 tests completed (96.3%)
[16:21:53] Progress: 389/404 tests completed (96.3%)
[16:22:23] Progress: 393/404 tests completed (97.3%)
[16:22:53] Progress: 393/404 tests completed (97.3%)
[16:23:23] Progress: 393/404 tests completed (97.3%)
[16:23:53] Progress: 396/404 tests completed (98.0%)
[16:24:23] Progress: 400/404 tests completed (99.0%)
SSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSS•SSSSSSSSSSS•SSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSS•SSSS•SSS•SSSSSSSSSSSSSSSSSSSSSSS•S•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSS••SS•SSSS•SSS•SSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSS•SS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•S•SSSSSSSSSSSSS•S•SS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS••SSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSS•SSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSS•SSS••SSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSS•SSS•SSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSS•SSSSSSSSSSSSSSSSSSSSSSSSS•SSSSSSSS•SS

Summarizing 3 Failures:
  [FAIL] [sig-architecture] Conformance Tests [It] should have at least two untainted nodes [Conformance] [sig-architecture, Conformance]
  k8s.io/kubernetes/test/e2e/architecture/conformance.go:45
  [FAIL] [sig-api-machinery] Servers with support for API chunking [It] should support continue listing from the last key if the original version has been compacted away, though the list is inconsistent [Slow] [Conformance] [sig-api-machinery, Slow, Conformance]
  k8s.io/kubernetes/test/e2e/apimachinery/chunking.go:202
  [FAIL] [sig-apps] Daemon set [Serial] [It] should rollback without unnecessary restarts [Conformance] [sig-apps, Serial, Conformance]
  k8s.io/kubernetes/test/e2e/apps/daemon_set.go:446

Ran 404 of 6605 Specs in 6293.100 seconds
FAIL! -- 401 Passed | 3 Failed | 0 Pending | 6201 Skipped
--- FAIL: TestE2E (6293.24s)
FAIL

Ginkgo ran 1 suite in 1h44m53.537715399s

Test Suite Failed
2025/04/02 16:24:29 Saving results at /tmp/results
2025/04/02 16:24:30 running command: exit status 1
16:24:31 INF Tests finished after 1h44m54s.
16:24:31 INF Downloading e2e.log to e2e.log...
16:24:31 INF Downloading junit_01.xml to junit_01.xml...
16:24:31 INF Waiting for Pod to terminate...
16:24:31 INF Container conformance-container terminated.
16:24:31 INF Deleted ClusterRoleBinding conformance-serviceaccount-role:conformance.
16:24:31 INF Deleted ClusterRole conformance-serviceaccount:conformance.
16:24:31 INF Waiting for Namespace conformance to be deleted.
16:25:13 INF Deleted Namespace conformance.
16:25:13 ERR Tests failed (code 1).

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
Sounds good to me, there will be more data transfer but this should be more robust.

**Comment by farazkhawaja** [comment]
thanks guys for the review. Ive addressed the review comments. PTAL

**Comment by BenTheElder** [comment]
I'm testing this out now, thanks! 30s feels pretty good, tests take 2h or so (YMMV) so we will give plenty of progress updates over that time frame.

**Comment by rjsadow** [comment]
Our conformance checks seem to be running ok with these changes.

Ref: https://testgrid.k8s.io/sig-testing-misc#hydrophone-conformance-check

**Comment by BenTheElder** [comment]
> Our conformance checks seem to be running ok with these changes.

yes, but those are on clusters that pass all tests.

For an edge case, I tried this on a GKE autopilot cluster with restricts many things required by the tests so we get failure logs

**Comment by farazkhawaja** [comment]
@BenTheElder thanks for the test. just trynna understand the behavior, what exactly is printing in a loop?

File: pkg/conformance/client/logs.go [hunk 1]
```diff
@@ -112,7 +123,116 @@ func (c *Client) streamPodLogs(ctx context.Context, stream streamLogs) {
...
+				})
+			if err != nil {
+				stream.doneCh <- true
+			}
+
+			// Parse test progress
+			output := strings.TrimSpace(stdout.String())
+			totalTests, completedTests, err := parseTestProgress(output)
+			if err != nil {
+				stream.errCh <- err
```
Comment (farazkhawaja) [inline pkg/conformance/client/logs.go:213]: @rjsadow this looks like the problem causing flaky test which was mentioned in the slack channel

Im thinking of the following 
```
output := strings.TrimSpace(stdout.String())
			if output == "" {
				// Log file not ready yet, wait and try again
				time.Sleep(c.configuration.ProgressStatusInterval)
				continue
			}
```
